test/FastTests/Corax/StreamingOptimization.cs

```diff
@@ -0,0 +1,224 @@
+using FastTests.Voron;
+using Newtonsoft.Json;
+using Raven.Client.Documents;
+using Raven.Client.Documents.Indexes;
+using Raven.Client.Documents.Queries;
+using Raven.Client.Documents.Session;
+using Raven.Client.Json.Serialization.NewtonsoftJson.Internal;
+using Raven.Server.Documents.Indexes.Persistence.Corax;
+using Raven.Server.Documents.Queries;
+using Sparrow.Json;
+using Sparrow.Server;
+using Sparrow.Threading;
+using Tests.Infrastructure;
+using Xunit;
+using Xunit.Abstractions;
+using Index = Raven.Server.Documents.Indexes.Index;
+
+namespace FastTests.Corax;
+
+public class StreamingOptimization(ITestOutputHelper output) : RavenTestBase(output)
+{
+    [Fact] // TermMatch and asc order on same field with same type => can optimize
+    public async Task SortingMatchIsSkippedOnSingleTermMatch() => await TestQueryBuilder<TermMatch>(session => 
+        session.Advanced.AsyncDocumentQuery<Dto, DtoIndex>()
+        .WhereEquals(p => p.Name, "maciej")
+        .OrderBy(x => x.Name)
+        .GetIndexQuery());
+   
+    [Fact] // TermMatch and desc order on same field with same type => can optimize
+    public async Task SortingMatchIsNotSkippedOnSingleTermMatchDesc() => await TestQueryBuilder<TermMatch>(session => 
+        session.Advanced.AsyncDocumentQuery<Dto, DtoIndex>()
+            .WhereEquals(p => p.Name, "maciej")
+            .OrderByDescending(x => x.Name)
+            .GetIndexQuery());
+
+    [Fact] // TermMatch and asc order on same field with different type => cant optimize
+    public async Task SortingMatchIsNotSkippedOnSingleTermMatchNumeric() => await TestQueryBuilder<SortingMatch>(session => 
+        session.Advanced.AsyncDocumentQuery<Dto, DtoIndex>()
+            .WhereEquals(p => p.Name, "maciej")
+            .OrderByDescending(x => x.Name, OrderingType.Long)
+            .GetIndexQuery());
+
+    [Fact] // TermMatch and desc order on same field with same type => can optimize
+    public async Task SortingMatchIsSkippedWhenWeQueryOnTheSameType() => await TestQueryBuilder<TermMatch>(session => 
+        session.Advanced.AsyncDocumentQuery<Dto, DtoIndex>()
+            .WhereEquals(p => p.First, (object)1L)
+            .OrderBy(x => x.First, OrderingType.Long)
+            .GetIndexQuery());
+    
+    [Fact] // TermMatch and desc order on same field with same type => can optimize
+    public async Task SortingMatchIsSkippedWhenWeQueryOnTheSameType2() => await TestQueryBuilder<TermMatch>(session => 
+        session.Advanced.AsyncDocumentQuery<Dto, DtoIndex>()
+            .WhereEquals(p => p.First, (object)1D)
+            .OrderBy(x => x.First, OrderingType.Double)
+            .GetIndexQuery());
+    
+    [Fact] //MultiTermMatch => cannot optimize
+    public async Task SortingMatchIsNotSkippedOnMultiTermMatch() => await TestQueryBuilder<SortingMatch>(session => 
+        session.Advanced.AsyncDocumentQuery<Dto, DtoIndex>()
+            .WhereLessThan(p => p.First, 10)
+            .OrderBy(x => x.First)
```

Comment:
Why can't we optimize this?
Logically speaking, we are going to be iterating the terms in order, no?

```
for entry in index.EntriesFor("First");
   if entry.Term >= 10:
        break
    yield entry
```

That should return them sorted, no?

Reply:
In general, that is probably a very important optimization, since it matches:

```
from Posts 
where PublishedAt < $lastWeek
order by PublishedAt desc
limit 5
```

Meaning give me the 5 latest posts from last week.

Reply:
We cannot do streaming there due to duplicates removal:
https://github.com/ravendb/ravendb/blob/b07807e20ef79ca80d4bc0649b036836d49e4af6/src/Corax/Queries/MultiTermMatch.cs#L104-L107

Reply:
Yes, we could modify this code but we've to thing about `BinaryMatch` too:
https://github.com/ravendb/ravendb/blob/0c066497d8a7061809ff5e5dd8e350b8c5aded6e/src/Corax/Queries/BinaryMatch.Boolean.cs#L51-L58

Reply:
What sort of query will trigger this?

Reply:
When inner has more items than our buffer. `AndWith` requires a sorted buffer as input.